R/get-path-from-object.R

```diff
@@ -158,45 +158,234 @@ get_model_id.bbi_model <- function(.mod) {
+      rlang::abort(
+        c(
+          "x" = "Input data file does not exist or cannot be opened",
+          "i" = glue("Referenced input data path: {data_path}")
+        )
+      )
+    }
+  }
+
+  return(data_path)
+}
+
+
+
+#' Get the data path from a control stream file
+#' @noRd
+get_data_path_from_ctl <- function(.mod, normalize = TRUE){
+  check_model_object(.mod, "bbi_nonmem_model")
+  mod_path <- get_model_path(.mod)
+  ctl <- nmrec::read_ctl(mod_path)
+
+  # Get data record
+  data_recs <- nmrec::select_records(ctl, "data")
+  n_data <- length(data_recs)
+  if(n_data !=1){
+    recs_fmt <- purrr::map_chr(data_recs, function(rec) rec$format())
+    rlang::abort(
+      c(
+        glue::glue("Expected a single data record, but found {n_data}:\n\n"),
+        glue("{recs_fmt}")
+      )
+    )
+  }
+
+  # Get file path and remove quotes
+  data_path_ctl <- nmrec::get_record_option(data_recs[[1]], "filename")$value
+  data_path_ctl <- unquote_filename(data_path_ctl)
+
+  if(isTRUE(normalize)){
+    # Normalize to current model directory
+    model_dir <- dirname(get_model_path(.mod, .check_exists = FALSE))
+    # Adjust for model extension
+    data_path_ctl_adj <- adjust_data_path_ext(data_path_ctl, mod_path)
+    data_path <- fs::path_norm(file.path(model_dir, data_path_ctl_adj)) %>%
+      as.character()
+  }else{
+    data_path <- data_path_ctl
+  }
+
+  return(data_path)
+}
+
+
+#' Get data path as defined in bbi_config.json
+#' @noRd
+get_data_path_from_json <- function(.mod, normalize = TRUE){
   cfg_path <- get_config_path(.mod, .check_exists = FALSE)
 
   if (!fs::file_exists(cfg_path)) {
-    stop(paste(
-      glue("Cannot extract data path because no config file exists as {cfg_path}"),
-      "This likely means the model has not successfully been run yet.",
-      sep = "\n"
-    ))
+    rlang::abort(
+      c(
+        glue("Cannot extract data path from config file because none exist at `{cfg_path}`"),
+        "i" = "This likely means the model has not successfully been run yet."
+      )
+    )
   }
 
   cfg <- jsonlite::fromJSON(cfg_path)
+  data_path_config <- cfg[[CONFIG_DATA_PATH]]
+
+  if(isTRUE(normalize)){
+    # Normalize to output directory
+    output_dir <- get_output_dir(.mod, .check_exists = FALSE)
+    data_path <- fs::path_norm(file.path(output_dir, data_path_config)) %>%
+      as.character()
+  }else{
+    data_path <- data_path_config
+  }
 
-  fs::path_norm(
-    file.path(
-      get_output_dir(.mod),
-      cfg[[CONFIG_DATA_PATH]]
-    )
-  ) %>%
-    as.character()
+  return(data_path)
+}
+
+#' Adjust the data path based on the model extension
+#'
+#' @param data_path a file path to an input data set
+#' @param mod_path a relative or absolute model path. Must include the extension.
+#'
+#' @noRd
+adjust_data_path_ext <- function(data_path, mod_path){
+  checkmate::assert_true(is_valid_nonmem_extension(mod_path))
+
+  if(grepl("ctl", fs::path_ext(mod_path))){
```

Comment:
nit-pick: I think the `grepl` use here is confusing because a reader needs to look at the callers to verify that this can only be "mod" or "ctl", so there's no need to worry about this pattern matching some mis-specified extension like "ctlo" or "actl".

In my opinion the `identical(fs::path_ext("mod_path"), "ctl")` I suggested <https://github.com/metrumresearchgroup/bbr/pull/664#discussion_r1499293489> would make things clearer.